data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,545 @@
+    /// Construct empty [`ConstString`].
+    #[inline]
+    pub const fn new() -> Self {
+        Self(ConstStringData::new())
+    }
+}
+
+impl Deref for ConstString {
+    type Target = str;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_ref()
+    }
+}
+
+impl<T: ?Sized> AsRef<T> for ConstString
+where
+    ConstStringData: AsRef<T>,
+{
+    #[inline]
+    fn as_ref(&self) -> &T {
+        self.0.as_ref()
+    }
+}
+
+impl<T: Into<ConstStringData>> From<T> for ConstString {
+    #[inline]
+    fn from(value: T) -> Self {
+        Self(value.into())
+    }
+}
+
+impl Display for ConstString {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        write!(f, "{}", &**self)
+    }
+}
+
+impl Debug for ConstString {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        write!(f, "{:?}", &**self)
+    }
+}
+
+impl Hash for ConstString {
+    #[inline]
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        (**self).hash(state)
+    }
+}
+
+impl PartialOrd for ConstString {
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        PartialOrd::partial_cmp(&**self, &**other)
+    }
+}
+
+impl Ord for ConstString {
+    #[inline]
+    fn cmp(&self, other: &Self) -> Ordering {
+        Ord::cmp(&**self, &**other)
+    }
+}
+
+impl PartialEq for ConstString {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        PartialEq::eq(&**self, &**other)
+    }
+}
+
+macro_rules! impl_eq {
+    ($($ty:ty,)*) => {
+        impl_eq!($($ty),*);
+    };
+    ($($ty:ty),*) => {$(
+        impl PartialEq<$ty> for ConstString {
+            #[allow(clippy::string_slice, clippy::deref_by_slicing)]
+            #[inline]
+            fn eq(&self, other: &$ty) -> bool {
+                PartialEq::eq(&self[..], &other[..])
+            }
+        }
+
+        impl PartialEq<ConstString> for $ty {
+            #[allow(clippy::string_slice, clippy::deref_by_slicing)]
+            #[inline]
+            fn eq(&self, other: &ConstString) -> bool {
+                PartialEq::eq(&self[..], &other[..])
+            }
+        }
+    )*};
+}
+
+impl_eq!(String, str, &str);
+
+impl Eq for ConstString {}
+
+impl Serialize for ConstString {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(self)
+    }
+}
+
+impl Encode for ConstString {
+    fn size_hint(&self) -> usize {
+        self.as_bytes().size_hint()
+    }
+
+    fn encode_to<W: Output + ?Sized>(&self, dest: &mut W) {
+        self.as_bytes().encode_to(dest)
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        self.as_bytes().encode()
+    }
+
+    fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
+        self.as_bytes().using_encoded(f)
+    }
+}
+
+impl IntoSchema for ConstString {
+    fn type_name() -> String {
+        String::type_name()
+    }
+    fn schema(map: &mut MetaMap) {
+        String::schema(map);
+    }
+}
```

Comment:
are `String` and `ConstString` byte representations exactly the same when serialized to Scale or Json?

Reply:
Yes, because in both cases structs are serialized as `str` under the hood.
I will change code to use [`WrapperTypeEncode`](https://docs.rs/parity-scale-codec/2.3.1/parity_scale_codec/trait.WrapperTypeEncode.html) to make it more clear and add test to compare serialization of `ConstString`  and `String`.